pkg/planner/core/tiflash_predicate_push_down.go

```diff
@@ -0,0 +1,371 @@
+			return true
+		case ast.LocalTimestamp, ast.MakeDate, ast.MakeTime, ast.MonthName, ast.PeriodAdd, ast.PeriodDiff, ast.Quarter, ast.SecToTime, ast.ToSeconds:
+			return true
+		case ast.StrToDate, ast.SubDate, ast.SubTime, ast.TimeLiteral, ast.TimeFormat, ast.TimeToSec, ast.TimeDiff, ast.TimestampDiff:
+			return true
+		// regexp functions
+		case ast.Regexp, ast.RegexpLike, ast.RegexpReplace, ast.RegexpSubstr, ast.RegexpInStr:
+			return true
+			// TODO: add more heavy cost functions
+		}
+	}
+	return false
+}
+
+// predicatePushDownToTableScan is used to push down the some filter conditions to the tablescan.
+// @param: sctx: the session context
+// @param: conds: the filter conditions
+// @param: ts: the PhysicalTableScan to be pushed down to
+func predicatePushDownToTableScan(sctx base.PlanContext, conds []expression.Expression, ts *PhysicalTableScan) {
+	// When the table is small, there is no need to push down the conditions.
+	if ts.tblColHists.RealtimeCount <= tiflashDataPackSize || ts.KeepOrder || len(conds) == 0 {
+		return
+	}
+
+	// group the conditions by columns and sort them by selectivity
+	sortedConds := groupByColumnsSortBySelectivity(sctx, conds, ts)
+
+	selectedConds := make([]expression.Expression, 0, len(conds))
+	selectedIncome := 0.0
+	selectedColumnCount := 0
+	selectedSelectivity := 1.0
+	totalColumnCount := len(ts.Columns)
+	tableRowCount := ts.StatsInfo().RowCount
+
+	for _, exprGroup := range sortedConds {
+		mergedConds := append(selectedConds, exprGroup.exprs...)
+		selectivity, _, err := cardinality.Selectivity(sctx, ts.tblColHists, mergedConds, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		colCnt := expression.ExtractColumnSet(mergedConds...).Len()
+		income := (1 - selectivity) * tableRowCount
+		// If selectedColumnCount does not change,
+		// or the increase of the number of filtered rows is greater than tiflashDataPackSize and the income increases, push down the conditions.
+		if colCnt == selectedColumnCount || (income > tiflashDataPackSize && income*(float64(totalColumnCount)-float64(colCnt)) > selectedIncome) {
+			selectedConds = mergedConds
+			selectedColumnCount = colCnt
+			selectedIncome = income * (float64(totalColumnCount) - float64(colCnt))
+			selectedSelectivity = selectivity
+		} else if income < tiflashDataPackSize {
+			// If the increase of the number of filtered rows is less than tiflashDataPackSize,
+			// break the loop to reduce the cost of calculating selectivity.
+			break
+		}
+	}
+
+	if len(selectedConds) == 0 {
+		return
+	}
+	logutil.BgLogger().Debug("planner: push down conditions to table scan", zap.String("table", ts.Table.Name.L), zap.String("conditions", string(expression.SortedExplainExpressionList(sctx.GetExprCtx().GetEvalCtx(), selectedConds))))
```

Comment:
In fact, Debug-level logs are rarely enabled and are not usually checked, so it is recommended to place them under the ```sctx.GetSessionVars().StmtCtx.AppendWarning()```.

Reply:
Is it your debug log? if it is ,it can be removed.